packages/replayio/src/utils/createDeferred.ts

```diff
@@ -0,0 +1,94 @@
+export const STATUS_NOT_FOUND = "not-found";
+export const STATUS_PENDING = "pending";
+export const STATUS_ABORTED = "aborted";
+export const STATUS_REJECTED = "rejected";
+export const STATUS_RESOLVED = "resolved";
+
+export type StatusNotFound = typeof STATUS_NOT_FOUND;
+export type StatusPending = typeof STATUS_PENDING;
+export type StatusAborted = typeof STATUS_ABORTED;
+export type StatusRejected = typeof STATUS_REJECTED;
+export type StatusResolved = typeof STATUS_RESOLVED;
+
+export interface Deferred<Type, Data = void> {
+  data: Data | undefined;
+  debugLabel: string | undefined;
+  promise: Promise<Type>;
+  rejection: Error | undefined;
+  reject(error: Error): void;
+  resolve(value?: Type): void;
+  resolution: Type | undefined;
+  status: StatusPending | StatusRejected | StatusResolved;
+}
+
+export type Status =
+  | StatusNotFound
+  | StatusPending
+  | StatusAborted
+  | StatusRejected
+  | StatusResolved;
+
+export function createDeferred<Type, Data = void>(
+  data?: Data,
+  debugLabel?: string
+): Deferred<Type, Data> {
+  let rejection: Error | undefined = undefined;
+  let resolution: Type | undefined = undefined;
+  let status: StatusPending | StatusRejected | StatusResolved = STATUS_PENDING;
+
+  let rejectPromise: (error: Error) => void;
+  let resolvePromise: (value: Type | PromiseLike<Type>) => void;
+
+  const promise = new Promise<Type>((resolve, reject) => {
+    rejectPromise = reject;
+    resolvePromise = resolve;
+  });
+  promise.catch(() => {
+    // Prevent unhandled promise rejection warning.
+  });
```

Comment:
Any particular reason why we want to prevent this here? I usually consider it a good thing that it warns/crashes the process when promises are unhandled. Fail fast and fix things and all of that.

Reply:
I think it's bad UX for common case– like for instance, if we're prompting you if you'd like to upload a new recording and you CTL+C quit, I don't think we should log a long error stack trace.

Reply:
Yeah, I wouldn't like this to be printed to the user. It just seems that maybe the fact that it's printed is a sign that some promise stays unhandled and that it should be fixed elsewhere?